### DISCUSSION TITLE
New builds Q2 2024

### DISCUSSION THREAD

**Discussion by ianw1974** [ianw1974/zimbra-build-scripts#81] [Announcements]
New builds went live on my site yesterday evening.  Tested both Zimbra 9.0.0 and Zimbra 10.0.7 for Ubuntu 20.04, so the remaining builds should be fine too.  I didn't see anything at least that wasn't working.  All services started and sending/receiving messages was possible.  These were also upgrade tests, from my previous builds.

A fix was made to disable ```libphp.so``` which is most likely something Zimbra is working on for the future.  For those who are long-time followers of my project will know I build from the ```develop``` branch, which means some additional patching is required, when we have newer changes that aren't supported or work as intended in previous versions of Zimbra.  Another example of this when nginx permissions were changed, a patch was made to address this for Zimbra 9.0.0 - although Zimbra 10.0.6 was not affected.

These will also be the last builds made for CentOS 7 and RHEL7, due to the fact they both go EOL before my Q3 builds are due.  I only provide builds for systems that actively supported.  This also means that Ubuntu 18.04 has also been dropped from my builds, since this has been EOL since April 2023 - although I built them far longer than I should have.

**Comment by jeredfloyd** [comment]
Thanks!

There were updates to the zimbra-1010-oss repo this morning, which are likely related to the` libphp.so` change you encountered.  The notable changes that will be picked up by the OS package manager are:

- postfix 3.6.1 -> 3.6.14
- php 7.4.27 -> 8.3.0
- zimbra-spell-components 2.0.13 -> 2.0.14

That's a pretty significant PHP version jump and seems like it may have code changes associated.  I'm a bit confused by Zimbra shipping PHP at all.  The only PHP seems to be the policyD admin UI (which is not available/supported) and aspell.  If I had to guess, there is an impending CVE fix related to aspell, but I haven't dug into it.

This suggests a 10.0.8 in the next few days.  Hopefully it does not have meaningful security content.

**Reply by jeredfloyd** [reply]
Ah, here is the patch showing the upgrade to PHP 8.  I think that your 10.0.7 packages will not have a functional spellchecker:
https://github.com/Zimbra/zm-aspell/commit/0e78284082d73e1b036a0321532376c354da755c

**Comment by ianw1974** [comment]
The update included ```zimbra-php``` which includes the ```libphp.so``` file.  So after an ```apt-get upgrade``` or whatever command to update on the distro you use, you can then do:

```
find /opt/zimbra -iname libphp.so
```

which will show it exists in ```/opt/zimbra/common/lib/apache2/modules/libphp.so```.  You can then edit ```/opt/zimbra/conf/httpd.conf``` and find the line with ```libphp.so``` and unhash it.  Then restart Zimbra.

This didn't exist before, hence it being commented out in my patch.  Obviously new builds now can have this patch removed, and will then use that module.  If 10.0.8 will be releases in the next few days, I'll just replace the builds with new ones.

**Reply by jeredfloyd** [reply]
From the Zimbra forums, it's been pointed out that this skew is probably due to you working from the `develop` branch, as you note above.  That branch includes a commit which adds the zimbra-1010-oss repo (https://github.com/Zimbra/zm-build/commit/e110e7b8fb3da430e06cd05d386ce1917e6e3b58) which is where the package updates have pulled from.  These are meant for the forthcoming 10.1 release.

**Reply by ianw1974** [reply]
Yes I know about that.  Previously for 9.0.0 and 10.0.x there were very little things to patch.  Obviously 10.1 is a bigger update, and they are preparing more things, so I will no doubt be patching more when things are only half-commited by Synacor.  Had they commited the libphp.so and also included the updates to PHP at the same time, things would have been fine.  But what they did was add libphp.so before it actually existed in a package.  Which doesn't make sense to me.  If you are going to include it, you put both things at the same time.  But patching and removing the config line is no big deal.  And it's resolved now anyway.

For those who want builds from tags, they can use the Maldua builds, which is posted about already on the forums.  It's also linked in the open issue here on my repo.  Maybe at some point I will offer to build from tags, but it's not a priority for me right now.

The way I build means that I have more than a standard 10.0.6 or 10.0.7 release, which can also mean extra bug fixes that the tag releases don't have.  But it also means there are minor issues/regressions like this to take into account and I patch to work around it.

People can choose to use or not use my builds, and I don't hide the fact that I build from the develop branch either.  I have been very open about it, both here and on the forums.

**Comment by ianw1974** [comment]
Here is proof that the spellchecker, after doing all that now works:

![brave_screenshot_zimbra upware pl](https://github.com/ianw1974/zimbra-build-scripts/assets/23740709/989512c7-0713-44a0-9165-0c3b32e85293)

**Comment by jeredfloyd** [comment]
Thanks!  Looks good to me.  For anyone who does this, it may be desirable to pull in the aspell.php changes from the linked commit as well.

As an aside -- can we talk about how wild it is that Zimbra pulls in all of Apache and all of PHP for... just the spellcheck endpoint?  Surely JavaSpell and a SOAP endpoint would be an easy modification, and save a whole lot of resources and attack surface.

**Reply by ianw1974** [reply]
I can only assume that Zimbra/Synacor plan perhaps to do more functionality with PHP in the future.  But they will have to keep up-to-date on it, otherwise we'll end up with another situation like how it was with Java in Zimbra - real old with a lot of critical bugs.  A lot of that has been addressed now.  They can generally stick with a certain release of PHP for about 2 years.  The company where I work, has a PHP application, and we moved to PHP 8.1 when it was a little after it was first released.  It meant 2 years peace and quiet before having to move to a new version.  We have subsequently moved to PHP 8.3, which means we are good for another 2 years now before having to upate again.

Providing Zimbra keep on top of it, to ensure whatever PHP code they have in Zimbra, that it will work/get updated regularly and not lagging behind, then shouldn't be a huge problem.

**Comment by jeredfloyd** [comment]
FFS -- Zimbra just pushed these package changes to the zimbra-1000-oss repo as part of the 10.0.8 patch.

I wonder how hard it would be to modify the build systems such that  zimbra packages get proper version dependencies on the OSS pages.  That way attempting to update, for example, zimbra-php would fail because the 10.0.7 packages depend on the current zimbra-php (and since the build packages aren't in a repo, the dependencies wouldn't resolve).

**Comment by ianw1974** [comment]
New builds going up shortly to replace the existing ones.  This means replacing 10.0.7 with 10.0.8 ones that no longer require patching for the libphp.so issue previously.  And so this patch has been removed from the build process.

**Comment by jeredfloyd** [comment]
Updating 10.0.6 -> 10.0.8 looks successful; thanks!

During the upgrade I encountered:
`Port conflict detected: 7780 (zimbra-spell)`

It looks like the shutdown phase of the install script did not stop the old zimbra-spell httpd properly, which is probably due to the underlying package already having been updated.  Rebooting the system resolved any issues.

**Comment by ianw1974** [comment]
Yep, most likely during upgrade, it normally does those port checks before starting the upgrade process.  Since Zimbra is running it says that - normally you can sometimes see ones mentioning DNS ports 53 sometime as well if resolvconf is running on a system for example.  Usually with that running on localhost port 53 it doesn't conflict so it's safe to continue.

Since zimbra-spell is part of Zimbra, it won't be problematic, since once the upgrade starts and before it finishes the services are restarted anyway.  A reboot in this instance wasn't even needed.  The port conflict would only be problematic, if you had additional applications running on your Zimbra server that would conflict with Zimbra.  In reality, the Zimbra spell port conflict shouldn't be reported if the process matches what is being installed/upgraded by Zimbra.  Had it been a different service, eg: aspell that isn't part of Zimbra on port 7780 then that would be a bigger problem in the sense the zimbra-spell service wouldn't start when one is already running.

Synacor/Zimbra should really make that detection work properly.

**Reply by jeredfloyd** [reply]
I really wish Synacor would make their repos open to PRs; they still haven't implemented some very serious iCal-processing issues that I provided solutions for months ago.  It's incredibly irritating that they use the excuse of "there's no community participation" to stop their FOSS builds, but then make it literally impossible for outside contributors to help.  Ah well, this is nothing new.

**Reply by ianw1974** [reply]
Yeah totally.  Bit of a catch 22 there.
